File: examples/FSharpDev/CustomReporting/CustomReporting.fs
module FSharpDev.CustomReportingExample

open System.Threading.Tasks
open FSharp.Control.Tasks.NonAffine

open NBomber
open NBomber.Contracts
open NBomber.Contracts.Stats
open NBomber.FSharp

let reportingSink = {
    new IReportingSink with
        member _.SinkName = "TestSink"
        member _.Init(context, infraConfig) = Task.CompletedTask
        member _.Start() = Task.CompletedTask
        member _.SaveRealtimeStats(stats) = Task.CompletedTask
        member _.SaveFinalStats(stats) = Task.CompletedTask
        member _.Stop() = Task.CompletedTask
        member _.Dispose() = ()
}

let run () =

    let step = Step.create("step", fun context -> task {
        do! Task.Delay(milliseconds 100)
        return Response.ok(sizeBytes = 100)
    })

    Scenario.create "simple_scenario" [step]
    |> Scenario.withoutWarmUp
    |> Scenario.withLoadSimulations [KeepConstant(copies = 1, during = minutes 1)]
    |> NBomberRunner.registerScenario
    |> NBomberRunner.withTestSuite "reporting"
    |> NBomberRunner.withTestName "custom_reporting_test"
    |> NBomberRunner.withReportingSinks [reportingSink]
    |> NBomberRunner.withReportingInterval(seconds 10)
    |> NBomberRunner.withReportFolder "./custom_reports"
    |> NBomberRunner.withReportFormats [ReportFormat.Html]
    |> NBomberRunner.run
    |> ignore

File: examples/CSharpDev/CustomReporting/CustomReporting.cs
using System;
using System.Threading.Tasks;

using Microsoft.Extensions.Configuration;
using Serilog;

using NBomber.Configuration;
using NBomber.Contracts;
using NBomber.Contracts.Stats;
using NBomber.CSharp;
using static NBomber.Time;

namespace CSharpDev.CustomReporting
{
    class CustomReportingSink : IReportingSink
    {
        private ILogger _logger;
        public string SinkName => nameof(CustomReportingSink);

        public Task Init(IBaseContext context, IConfiguration infraConfig)
        {
            _logger = context.Logger;
            return Task.CompletedTask;
        }

        public Task Start() => Task.CompletedTask;
        public Task SaveRealtimeStats(ScenarioStats[] stats) => Task.CompletedTask;
        public Task SaveFinalStats(NodeStats[] stats) => Task.CompletedTask;
        public Task Stop() => Task.CompletedTask;

        public void Dispose()
        { }
    }

    public class CustomReporting
    {
        public static void Run()
        {
            var step = Step.Create("step", async context =>
            {
                await Task.Delay(Seconds(0.1));
                return Response.Ok(sizeBytes: 100);
            });

            var scenario = ScenarioBuilder
                .CreateScenario("simple_scenario", step)
                .WithoutWarmUp()
                .WithLoadSimulations(Simulation.KeepConstant(1, TimeSpan.FromMinutes(1)));

            NBomberRunner
                .RegisterScenarios(scenario)
                .WithTestSuite("reporting")
                .WithTestName("custom_reporting_test")
                .WithReportingSinks(new CustomReportingSink())
                .WithReportingInterval(Seconds(10))
                .WithReportFolder("./custom_reports")
                .WithReportFormats(ReportFormat.Html, ReportFormat.Txt, ReportFormat.Csv)
                .Run();
        }
    }
}
